### ISSUE TITLE
Jolokia doesn't timeout connections in proxy mode

### ISSUE THREAD

**Issue by artur-fijalkowski** [rhuss/jolokia#222]
In my organization we are running Jolokia as proxy on Tomcat 7 on Java 1.7.0 u80 (Oracle) on Debian 7.

We observed that one of the monitored servers (also Tomcat) is making all connections from Jolokia in permantent ESTABLISHED state (from TCP/IP perpsective).

This leads to exhaustion of number of available threads inside JVM as Jolokia never (or very slowly) closes them down.

**Comment by rhuss** [comment]
Which version of Jolokia are you using ? There was a fix introduced in 1.3.0 for a bug which didn't close the JSR-160 JMX connections in certain circumstances (see #170)

**Comment by artur-fijalkowski** [comment]
I guess this 1.3.2 version of Jolokia

{"timestamp":1445261552,"status":200,"request":{"type":"version"},"value":{"protocol":"7.2","config":{"detectorOptions":"{}","canonicalNaming":"true","maxCollectionSize":"0","includeStackTrace":"true","historyMaxEntries":"10","agentId":"10.16.20.37-53482-7c165ec0-servlet","debug":"false","dispatcherClasses":"org.jolokia.jsr160.Jsr160RequestDispatcher","maxDepth":"15","maxObjects":"0","discoveryEnabled":"false","serializeException":"false","agentType":"servlet","debugMaxEntries":"100"},"agent":"1.3.2","info":{"product":"tomcat","vendor":"Apache","version":"7.0.28"}}}

**Comment by rhuss** [comment]
How many queries do you against the Jolokia proxy in parallel ? For each Jolokia request a JSR-160 RMI connection is created afresh (which in turn uses one or more TCP connections, that's out of scope of Jolokia but plain JMX remote as defined in JSR-160), so you have at least that many TCP connections open. 

Do you see any errors in the Tomcat log acting as proxy ?

**Comment by artur-fijalkowski** [comment]
The problem is that every query to one particular host (triggered by check_jmx4perl nagios plugin) is establishing connection for JSR-160 RMI which is permanent. Plugin timeouts after 15 seconds, but  I see from netstat that connection to monitored server is still in ESTABLISHED state.

tcp        0      0 XXX-jol-002.:35030 YYY-adm-001.h:9004 ESTABLISHED 13559/java

This connection is not going to terminate at all... I guess this is because tomcat on YYY-adm-001 is never confirming it send all the data...

**Comment by rhuss** [comment]
This is the relevant code: https://github.com/rhuss/jolokia/blob/master/agent/jsr160/src/main/java/org/jolokia/jsr160/Jsr160RequestDispatcher.java#L78-L91

JSR-160 RMI is a request-response based protocol. So, Jolokia does a request (i.e. queries the MBeanServer) and when it received the answer, closes the connection. So in your case it seems that Jolokia never gets back any answer which e.g. happens when there is a firewall in between. This also explains why your Nagios plugins run in a timeout, too.

But that hanging request can have other issues. In order to help for the connection problem one could implement an timeout oneself (with a watchdog thread), but this probably won't solve your real connection problem to the RMI backend. 

You know, RMI is quite a beast when it comes to firewall. There's an initial handshake phase over a well known port where both parties negotiate an *arbitrary* port over which the rest of the communications go.  So there is usual no way to get this through firewalls.

Could it be that your network topology might be the reason for the hanging requests ?

**Comment by pmcnally-mkto** [comment]
We are experiencing a similar issue as described here where the jolokia war is deployed in proxy mode on Tomcat 6 and JDK 1.7.0 but there are no firewalls involved.  We were originally using ver 1.2.3 which I updated on Friday to 1.3.2 after reading you post, #170 but the problem persists.

The proxy is not getting back a response from a number of hosts and over time these open connections build up and sooner or later cause the tomcat to fail due to no more available connections.

There are many reasons why the RMI server might not respond, OOM, bad routing, etc.  The monitoring server will detect and handle these timeouts.  The proxy service should be configurable to timeout connections and not cause the rest of the healthy monitored servers to start alerting.

Can you add a timeout property to the jolokia proxy mode?  e.g. after X seconds close the connection, or set the value to 0 to keep it as it is now. This way the individual deployments and choose what best suits their environment.

**Comment by rhuss** [comment]
Unfortunately there are not many configuration hooks for tuning the RMI connection used by `MBeanServerConnection` doing remote calls. 

The only thing I found it so use a global property which affects all RMI communication. For Tomcat one could set 

```
CATALINA_OPTS=-Dsun.rmi.transport.tcp.responseTimeout=60000
```

in `bin/setenv.sh`  for a 60s timeout.

**Comment by pmcnally-mkto** [comment]
Hi Roland,

We appreciate the fast response.  I have added that parameter to the JVM and so far it looks like is doing the trick.  I will continue to keep any eye on the proxy and look for knock on effects but I think this solves the immediate problem.  We will be leaving the underlying issues to the development team to solve those but this will allow our monitoring to continue.

Many thanks,

Paul

**Comment by grgrzybek** [comment]
Closing this one, as `-Dsun.rmi.transport.tcp.responseTimeout=60000` did the trick.
The other improvement in JSR-160 area is:
* #401
